tensorboard/plugins/custom_scalar/custom_scalars_plugin.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""The TensorBoard Custom Scalars plugin.
+
+This plugin lets the user create scalars plots with custom run-tag combinations
+by specifying regular expressions.
+
+See `http_api.md` in this directory for specifications of the routes for
+this plugin.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import re
+import threading
+import time
+
+from google.protobuf import json_format
+from google.protobuf import text_format
+import tensorflow as tf
+from werkzeug import wrappers
+
+from tensorboard.backend import http_util
+from tensorboard.plugins import base_plugin
+from tensorboard.plugins.custom_scalar import layout
+from tensorboard.plugins.custom_scalar import scalars_layout_pb2
+from tensorboard.plugins.scalar import metadata as scalars_metadata
+from tensorboard.plugins.scalar import scalars_plugin
+
+# The name of the property in the response for whether the regex is valid.
+_REGEX_VALID_PROPERTY = 'regex_valid'
+
+# The name of the property in the response for the payload (tag to ScalarEvents
+# mapping).
+_TAG_TO_EVENTS_PROPERTY = 'tag_to_events'
+
+# The number of seconds to wait in between checks for the config file specifying
+# layout.
+_CONFIG_FILE_CHECK_THROTTLE = 60
+
+
+class CustomScalarsPlugin(base_plugin.TBPlugin):
+  """CustomScalars Plugin for TensorBoard."""
+
+  plugin_name = 'custom_scalars'
+
+  def __init__(self, context):
+    """Instantiates ScalarsPlugin via TensorBoard core.
+
+    Args:
+      context: A base_plugin.TBContext instance.
+    """
+    self._logdir = context.logdir
+    self._multiplexer = context.multiplexer
+    self._scalars_plugin = scalars_plugin.ScalarsPlugin(context)
```

Comment:
Just a note: [the distributions plugin similarly instantiates a histograms plugin](https://github.com/tensorflow/tensorboard/blob/f132640e6e04194581474ace09fe526eac7d697b/tensorboard/plugins/distribution/distributions_plugin.py#L50). At the time, @dandelionmane and I agreed that this was a Good Thing because they depend on the same data source and this code makes the data dependency especially explicit.